# DS-SF-34 | 02 | The `pandas` Library | Assignment | Starter Code

## The Bistro Meets `pandas`

You've just told one of your friend that you are taking a Data Science class.  (Yeah!)  Your friend is running a bistro, a small restaurant, serving moderately priced simple meals in a modest setting ([Wikipedia](https://en.wikipedia.org/wiki/Bistro)).  She collected over some period of time the following information of her patrons' visits.

| Variable's name | Its meaning |
|:---:|:---|
| `name` | Patron's first name |
| `gender` | Patron's gender |
| `is_smoker` | Whether the patron is smoking or not |
| `party` | Party's size |
| `check` | Check amount (\$) (after taxes but before tips) |
| `tip` | Tip (\$) that the patron added to the check |
| `day` | Week day of the visit |
| `time` | Rough time estimate of the visit |

In this assignment, we will be exploring this dataset using `pandas`.<sup>(*)</sup>

<sup>(*)</sup> this dataset was adapted from the `tips` dataset of the `seaborn` package (https://github.com/mwaskom/seaborn-data)

> ### Question 1.  Import `numpy` (as `np`) and `pandas` (as `pd`).

In [1]:
import numpy as np
import pandas as pd
import os

# TODO

pd.set_option('display.max_rows', 10)
pd.set_option('display.notebook_repr_html', True)
pd.set_option('display.max_columns', 10)

> ### Question 2.  Read the `dataset-02-tips.csv` dataset.

In [2]:
tips = pd.read_csv("../datasets/dataset-02-bistro.csv")
tips.head()

day    time      name  gender  is_smoker  party  check   tip
0  Sunday  Dinner  Kimberly  Female      False      2  16.99  1.01
1  Sunday  Dinner  Nicholas    Male      False      3  10.34  1.66
2  Sunday  Dinner     Larry    Male      False      3  21.01  3.50
3  Sunday  Dinner    Joseph    Male      False      2  23.68  3.31
4  Sunday  Dinner    Janice  Female      False      4  24.59  3.61

> ### Question 3.  What is the class of the `pandas` object storing the dataset?

In [3]:
type(tips)

pandas.core.frame.DataFrame

Answer: DataFrame

> ### Question 4.  How many samples (i.e., rows) are in this dataset?

In [4]:
tips.shape

(244, 8)

In [5]:
tips.shape[0]

244

Answer: 244

> ### Question 5.  How many variables (i.e., columns) are in this dataset?

In [6]:
tips.shape[1]

8

Answer: 8

> ### Question 6.  Print the name of each column in the dataset, one name per line.

In [7]:
for col in tips.columns:
    print col

day
time
name
gender
is_smoker
party
check
tip


In [8]:
list(tips.columns)

['day', 'time', 'name', 'gender', 'is_smoker', 'party', 'check', 'tip']

> ### Question 7.  Print the first two rows of the dataset to the console.  What does the output look like?

In [9]:
print tips[0:2]

      day    time      name  gender  is_smoker  party  check   tip
0  Sunday  Dinner  Kimberly  Female      False      2  16.99  1.01
1  Sunday  Dinner  Nicholas    Male      False      3  10.34  1.66


Answer: Output looks raw

> ### Question 8.  Extract the last 2 rows of the data frame and print them to the console.  What does the output look like?

In [10]:
print tips.tail(2)

          day    time    name  gender  is_smoker  party  check   tip
242  Saturday  Dinner     Jon    Male      False      2  17.82  1.75
243  Thursday  Dinner  Brandi  Female      False      2  18.78  3.00


Answer: Not good

In [11]:
tips.tail(2) # Much better

day    time    name  gender  is_smoker  party  check   tip
242  Saturday  Dinner     Jon    Male      False      2  17.82  1.75
243  Thursday  Dinner  Brandi  Female      False      2  18.78  3.00

> ### Question 9.  Does the dataset contain any missing values?

In [12]:
tips.isnull().sum()

day          0
time         0
name         0
gender       0
is_smoker    0
party        0
check        0
tip          0
dtype: int64

Answer: Don't think so

In [13]:
tips.describe()

party       check         tip
count  244.000000  244.000000  244.000000
mean     2.569672   19.785943    2.998279
std      0.951100    8.902412    1.383638
min      1.000000    3.070000    1.000000
25%      2.000000   13.347500    2.000000
50%      2.000000   17.795000    2.900000
75%      3.000000   24.127500    3.562500
max      6.000000   50.810000   10.000000

> ### Question 10.  What can you say about the `is_smoker` variable?  I.e., will it bring any insights when analyzing the dataset?  What do you want to do with it?  (and do it...)

In [14]:
tips.is_smoker.unique()

array([False], dtype=bool)

Answer: They are all false

In [15]:
tips.drop("is_smoker", axis = 1, inplace = True)

tips.head()

day    time      name  gender  party  check   tip
0  Sunday  Dinner  Kimberly  Female      2  16.99  1.01
1  Sunday  Dinner  Nicholas    Male      3  10.34  1.66
2  Sunday  Dinner     Larry    Male      3  21.01  3.50
3  Sunday  Dinner    Joseph    Male      2  23.68  3.31
4  Sunday  Dinner    Janice  Female      4  24.59  3.61

> ### Question 11.  For which week days does the dataset has data for?

In [16]:
tips.day.unique()

array(['Sunday', 'Saturday', 'Thursday', 'Friday'], dtype=object)

Answer: Not M, T, W

> ### Question 12.  How often was the bistro patronized for each week day?

(check `.value_counts()`; it could come in handy)

(http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.value_counts.html)

In [17]:
tips.day.value_counts()

Saturday    87
Sunday      76
Thursday    62
Friday      19
Name: day, dtype: int64

Answer:

> ### Question 13.  How much tip did waiters collect for each week day?

In [18]:
tips.groupby(['name', 'day'])['tip'].sum() #NOT ANSWER, BUT THE SYNTAX IS SOLID

name       day     
Adalberto  Friday      1.50
Aiden      Sunday      2.24
Alejandro  Saturday    3.27
           Sunday      3.18
Alex       Friday      1.50
                       ... 
Wanda      Sunday      5.20
Wilburn    Sunday      1.96
William    Saturday    6.05
Willie     Saturday    2.00
Zackary    Thursday    2.00
Name: tip, Length: 219, dtype: float64

In [19]:
tips.groupby('day')['tip'].sum()

day
Friday       51.96
Saturday    260.40
Sunday      247.39
Thursday    171.83
Name: tip, dtype: float64

 Answer: TODO

> ### Question 14.  What is the average tip per check (in absolute \$) for each week day?

In [20]:
tips.groupby('day')['tip'].mean()

day
Friday      2.734737
Saturday    2.993103
Sunday      3.255132
Thursday    2.771452
Name: tip, dtype: float64

Answer: TODO

> ### Question 15.  What is the average tip per check (as a percentage of the check) for each week day?

In [21]:
# Calculate tip % var
tips['tip_percent'] = tips.tip / tips.check

tips.head()

day    time      name  gender  party  check   tip  tip_percent
0  Sunday  Dinner  Kimberly  Female      2  16.99  1.01     0.059447
1  Sunday  Dinner  Nicholas    Male      3  10.34  1.66     0.160542
2  Sunday  Dinner     Larry    Male      3  21.01  3.50     0.166587
3  Sunday  Dinner    Joseph    Male      2  23.68  3.31     0.139780
4  Sunday  Dinner    Janice  Female      4  24.59  3.61     0.146808

In [22]:
tips.groupby(['day'])['tip_percent'].mean()

day
Friday      0.169913
Saturday    0.153152
Sunday      0.166897
Thursday    0.161276
Name: tip_percent, dtype: float64

Answer: TODO

> ### Question 16.  Are there any name in common between male and female patrons?  (E.g., `Chris` can refer to either a man or a woman)

(check `numpy.intersect1d()`; it could come in handy)

(https://docs.scipy.org/doc/numpy/reference/generated/numpy.intersect1d.html)

In [23]:
# Split into male and female dfs
df_male = tips[tips.gender == "Male"]
df_female = tips[tips.gender == "Female"]

df_male.head()

day    time      name gender  party  check   tip  tip_percent
1  Sunday  Dinner  Nicholas   Male      3  10.34  1.66     0.160542
2  Sunday  Dinner     Larry   Male      3  21.01  3.50     0.166587
3  Sunday  Dinner    Joseph   Male      2  23.68  3.31     0.139780
5  Sunday  Dinner      Mark   Male      4  25.29  4.71     0.186240
6  Sunday  Dinner     Jamie   Male      2   8.77  2.00     0.228050

In [24]:
np.intersect1d(df_male.name, df_female.name)

array(['Casey'], dtype=object)

Answer: Yep, Casey

> ### Question 17.  If no patrons share the same name, how many unique patrons are in the dataset?

In [25]:
len(tips.name.unique())

181

Answer: 182 (181 + the one intersection)

> ### Question 18.  How many times did `Kevin` patronized the bistro?  How about `Alice`?

In [26]:
tips[tips.name == "Kevin"]

day    time   name gender  party  check   tip  tip_percent
96     Friday  Dinner  Kevin   Male      2  27.28  4.00     0.146628
187    Sunday  Dinner  Kevin   Male      5  30.46  2.00     0.065660
218  Saturday  Dinner  Kevin   Male      2   7.74  1.44     0.186047
239  Saturday  Dinner  Kevin   Male      3  29.03  5.92     0.203927

In [27]:
tips[tips.name == "Kevin"].shape[0] #remember shape[0] is numrows, shape[1] is numcols

4

In [28]:
tips[tips.name == "Alice"].shape[0]

2

Answer: Kevin - 4
        Alice - 2

> ### Question 19.  Who are the top 3 female and male patrons?

In [34]:
df_male.name.value_counts()

David      8
James      5
Casey      5
Carl       4
Kevin      4
          ..
Curtis     1
Andrew     1
Mario      1
Ronald     1
Richard    1
Name: name, Length: 110, dtype: int64

In [35]:
df_female.name.value_counts()

Mary         4
Casey        3
Laura        3
Theresa      2
Alice        2
            ..
Paisley      1
Marian       1
Shaquasia    1
Wanda        1
Sarah        1
Name: name, Length: 72, dtype: int64

Answer: TODO

> ### Question 20.  Who's the best tipper (as a fraction of all tips over all check totals)?  Who's the worst?  How many times did they patronize the bistro?

(check `numpy.intersect1d()`; it could come in handy)

- (http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.sort_values.html)
- (http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.sort_values.html)

In [40]:
name_tip = tips.groupby(['name'])['tip_percent'].mean()

name_tip.sort_values()

name
Jeremy      0.035638
Mildred     0.056433
Destiny     0.056797
Kimberly    0.059447
Willie      0.066534
              ...   
Zackary     0.266312
Dennis      0.280535
Bailey      0.325733
Maryann     0.416667
Daniel      0.430215
Name: tip_percent, Length: 181, dtype: float64

In [41]:
name_tip

name
Adalberto    0.120385
Aiden        0.107900
Alejandro    0.141879
Alex         0.149199
Alexander    0.194837
               ...   
Wanda        0.149382
Wilburn      0.130319
William      0.195786
Willie       0.066534
Zackary      0.266312
Name: tip_percent, Length: 181, dtype: float64

In [46]:
# Frequency df
name_tip = name_tip.to_frame()
name_tip

tip_percent
name                  
Adalberto     0.120385
Aiden         0.107900
Alejandro     0.141879
Alex          0.149199
Alexander     0.194837
...                ...
Wanda         0.149382
Wilburn       0.130319
William       0.195786
Willie        0.066534
Zackary       0.266312

[181 rows x 1 columns]

In [61]:
# tips = tips.set_index('name')
# tips.index

tip2 = pd.read_csv("../datasets/dataset-02-bistro.csv")

In [63]:
# This somehow worked w/ one having name as index

pre = tip2.join(name_tip, on = 'name')

pre.head()

day    time      name  gender  is_smoker  party  check   tip  \
0  Sunday  Dinner  Kimberly  Female      False      2  16.99  1.01   
1  Sunday  Dinner  Nicholas    Male      False      3  10.34  1.66   
2  Sunday  Dinner     Larry    Male      False      3  21.01  3.50   
3  Sunday  Dinner    Joseph    Male      False      2  23.68  3.31   
4  Sunday  Dinner    Janice  Female      False      4  24.59  3.61   

   tip_percent  
0     0.059447  
1     0.160542  
2     0.166587  
3     0.111711  
4     0.174365

In [64]:
pre.shape

(244, 9)

In [94]:
name_freq = tip2.name.value_counts()
name_freq = name_freq.to_frame()
name_freq.index

Index([u'Casey', u'David', u'James', u'Mary', u'Carl', u'Kevin', u'Laura',
       u'Jonathan', u'Tyler', u'Kenneth',
       ...
       u'Curtis', u'Dennis', u'Paisley', u'Jacqueline', u'Vernon', u'Walter',
       u'Chandler', u'Reyes', u'Matthew', u'Sarah'],
      dtype='object', length=181)

In [73]:
name_freq.columns = ['visits']
name_freq

visits
Casey          8
David          8
James          5
Mary           4
Carl           4
...          ...
Walter         1
Chandler       1
Reyes          1
Matthew        1
Sarah          1

[181 rows x 1 columns]

In [98]:
post = pre
post.set_index('name')

post = name_freq.join(post, on = 'name', how = 'left')

ValueError: columns overlap but no suffix specified: Index([u'name'], dtype='object')

In [87]:
post.head()

visitspost  day time name gender     ...     party  check  tip  \
Casey           8  NaN  NaN  NaN    NaN     ...       NaN    NaN  NaN   
David           8  NaN  NaN  NaN    NaN     ...       NaN    NaN  NaN   
James           5  NaN  NaN  NaN    NaN     ...       NaN    NaN  NaN   
Mary            4  NaN  NaN  NaN    NaN     ...       NaN    NaN  NaN   
Carl            4  NaN  NaN  NaN    NaN     ...       NaN    NaN  NaN   

       tip_percent  visitsfreq  
Casey          NaN         NaN  
David          NaN         NaN  
James          NaN         NaN  
Mary           NaN         NaN  
Carl           NaN         NaN  

[5 rows x 11 columns]

Answer: TODO